temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+) -> Optional[datetime]:
+    if ts.seconds == 0 and ts.nanos == 0:
+        return None
+    # Protobuf doesn't set the timezone but we want to
+    return ts.ToDatetime().replace(tzinfo=timezone.utc)
+
+
+class SharedStateManager(ABC):
+    """Base class for a shared state manager providing cross-process-safe
+    primitives for use by activity executors.
+
+    Cross-worker use of the shared state manager is encouraged.
+    :py:meth:`create_from_multiprocessing` provides the commonly used
+    implementation.
+    """
+
+    # TODO(cretz): Document that queue should be a multiprocessing.Manager().Queue()
+    # and that executor defaults as a single-worker thread pool executor
+    @staticmethod
+    def create_from_multiprocessing(
+        mgr: multiprocessing.managers.SyncManager,
+        queue_poller_executor: Optional[concurrent.futures.Executor] = None,
+    ) -> SharedStateManager:
+        """Create a shared state manager from a multiprocessing manager.
+
+        Args:
+            mgr: Sync manager to create primitives from. This is usually
+                :py:func:`multiprocessing.Manager`.
+            queue_poller_executor: The executor used when running the
+                synchronous heartbeat queue poller. This should be a
+                :py:class:`concurrent.futures.ThreadPoolExecutor`. If unset, a
+                thread pool executor is created with max-workers of 1.
+
+        Returns:
+            The shared state manager.
+        """
+        return _MultiprocessingSharedStateManager(
+            mgr, queue_poller_executor or concurrent.futures.ThreadPoolExecutor(1)
+        )
+
+    @abstractmethod
+    def new_event(self) -> threading.Event:
+        """Create a threading.Event that can be used across processes."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def register_heartbeater(
+        self, task_token: bytes, heartbeat: Callable[..., None]
+    ) -> SharedHeartbeatSender:
+        """Register a heartbeat function.
+
+        Args:
+            task_token: Unique task token for the heartbeater.
+            heartbeat: Function that should be called when the resulting sender
+                is sent a heartbeat.
+
+        Returns:
+            A sender that can be pickled for use in another process.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def unregister_heartbeater(self, task_token: bytes) -> None:
+        """Unregisters a previously registered heartbeater for the task
+        token.
+        """
+        raise NotImplementedError
+
+
+class SharedHeartbeatSender(ABC):
+    """Base class for a heartbeat sender that is picklable for use in another
+    process.
+    """
+
+    @abstractmethod
+    def send_heartbeat(self, task_token: bytes, *details: Any) -> None:
+        """Send a heartbeat for the given task token and details."""
+        raise NotImplementedError
+
+
+class _MultiprocessingSharedStateManager(SharedStateManager):
```

Comment:
Ya, seems like if we didn't have to worry about activities in other processes that could eliminate a bunch of stuff. :man_shrugging: 

I can also see though why it's nice for users to not have to spawn inside activities themselves, though... So perhaps it's worth it.

Reply:
It doesn't eliminate much and I think this is a really common async approach in Python. And it's basically already done, heh. I think we can expect it to be heavily used.